### PR TITLE
Adding the CheMeleon foundation model instead of out-of-the-box chemprop models

### PR DESCRIPTION
**Pull request by akshatzalte**
Replace Chemprop D-MPNN with CheMeleon foundation model                                                                                
                                                                                                                                         
  Swaps the second model backend from a from-scratch Chemprop D-MPNN to CheMeleon, a pre-trained BondMessagePassing foundation model
  fine-tuned per endpoint.                                                                                                               
  
  Changes:                                                                                                                               
  - chemprop_utils.py: adds train_chemeleon — auto-downloads weights from Zenodo on first call, same protonation/log-transform protocol
  as XGBoost                                                                                                                             
  - All notebooks 2.07–2.15: --model chemprop → --model chemeleon
  - scripts/: adds train_xgboost.sh, train_chemeleon.sh, combine.sh, exploration.sh for running full pipelines with dependency ordering  
  and per-job logging

### CODE REVIEW

**Comment by akshatzalte** [comment]
# CheMeleon Integration: Implementation and Results

## Background

CheMeleon is a pre-trained molecular foundation model based on the BondMessagePassing (D-MPNN) architecture, introduced in Burns, Zalte, et al. [1]. Rather than training a graph neural network from scratch, CheMeleon fine-tunes a set of pre-trained message-passing weights that have been learned across a broad corpus of molecular data — the expectation being that these weights encode transferable chemical representations that generalize better than task-specific models trained from scratch on limited ADMET data.

## Implementation

The implementation lives in `src/polaris_generalization/chemprop_utils.py` (`train_chemeleon`).

**Pre-trained weights.** Foundation model weights (`chemeleon_mp.pt`) are downloaded automatically from Zenodo (record 15460715, DOI: `10.5281/zenodo.15426600`) on first use and cached at `data/interim/chemeleon_mp.pt`. The `BondMessagePassing` module is initialized with the checkpoint's hyperparameters and its weights are frozen/fine-tuned during training.

**Architecture.** A two-layer FFN with hidden dimension 1024 is appended on top of the pre-trained message-passing backbone:

```
BondMessagePassing (pre-trained, loaded from checkpoint)
    → MeanAggregation
    → RegressionFFN (2 layers × 1024 hidden, trained from scratch)
```

**Pre-processing.** The pre-processing pipeline is identical to XGBoost: molecules are protonated at the assay-relevant pH (7.4 for most endpoints, 6.5 for Caco-2) using `dimorphite_dl` before featurization. The same log-transform protocol (`log10(clip(x, 1e-10) + 1)`) is applied to all endpoints except LogD.

**Training.** A random 10% of the training fold is held out for early stopping (minimum 5 validation samples). No ensembling is used (single model per fold/endpoint). Predictions are cached as `.npy` arrays in `data/processed/{notebook}/chemeleon/pred_cache/` so re-runs skip retraining.

---

## Baseline Performance (Notebook 2.08)

CheMeleon is evaluated under the same train/test split used for XGBoost. Results are reported on the held-out test set.

| Endpoint | CheMeleon MAE | XGBoost MAE | CheMeleon R² | XGBoost R² | CheMeleon Spearman | XGBoost Spearman |
|---|---|---|---|---|---|---|
| LogD | **0.397** | 0.509 | **0.720** | 0.592 | **0.872** | 0.758 |
| KSOL | **0.341** | 0.475 | **0.462** | 0.295 | **0.538** | 0.500 |
| HLM CLint | **0.294** | 0.314 | **0.328** | 0.231 | **0.630** | 0.562 |
| MLM CLint | 0.396 | **0.389** | 0.170 | **0.189** | **0.475** | 0.464 |
| Caco-2 Papp A>B | **0.297** | 0.308 | **0.206** | 0.192 | **0.539** | 0.512 |
| Caco-2 Efflux | **0.352** | 0.365 | **0.207** | 0.186 | **0.683** | 0.645 |
| MPPB | **0.216** | 0.282 | **0.533** | 0.223 | **0.754** | 0.616 |
| MBPB | **0.177** | 0.205 | **0.628** | 0.570 | **0.830** | 0.778 |
| MGMB | **0.220** | 0.250 | **0.463** | 0.452 | **0.716** | 0.687 |

CheMeleon outperforms XGBoost on nearly every endpoint and metric. Improvements are largest for LogD (Spearman +0.11, R² +0.13), KSOL (MAE −0.13), and protein binding (MPPB R² +0.31, MBPB MAE −0.03). The only endpoint where XGBoost has a slight edge is MLM CLint (MAE 0.389 vs 0.396, R² 0.189 vs 0.170) — a high-noise endpoint with limited training data.

---

## Performance-Over-Distance (Notebook 2.07)

Both models are evaluated on held-out molecules binned by Tanimoto distance from the training set (cluster, time, and target splits). The degradation ratio (RMSE in the farthest bin / RMSE in the closest bin) measures how steeply performance drops at the distribution boundary.

| Endpoint | Strategy | CheMeleon deg. ratio | XGBoost deg. ratio |
|---|---|---|---|
| LogD | cluster | 3.54 | 3.21 |
| LogD | time | 2.84 | 2.59 |
| HLM CLint | cluster | 4.30 | 2.98 |
| HLM CLint | time | 2.34 | 1.56 |
| KSOL | cluster | 1.34 | 1.90 |
| KSOL | time | 1.62 | 1.81 |
| MBPB | cluster | 2.54 | 1.82 |
| MPPB | cluster | 2.63 | 2.14 |

The qualitative trend is consistent across both models: performance degrades as test molecules move farther from training data, with the steepest degradation under cluster and time splits. CheMeleon shows slightly higher degradation ratios for several endpoints (HLM, MBPB, MPPB), suggesting that even though it starts from a lower absolute error, it is not uniformly more robust to distribution shift. For KSOL under cluster split, CheMeleon is more robust (1.34 vs 1.90). The overall pattern — substantial degradation in all split strategies except target split — is a finding that holds regardless of model architecture.

---

## IID vs OOD Across Chemical Series (Notebook 2.09)

The IID/OOD split separates molecules by chemical series: IID molecules belong to series seen in training; OOD molecules come from entirely new series.

| Endpoint | CheMeleon IID R² | CheMeleon OOD R² | XGBoost IID R² | XGBoost OOD R² |
|---|---|---|---|---|
| LogD | 0.640 | **0.444** | 0.645 | 0.055 |
| KSOL | 0.157 | −0.560 | 0.291 | **0.119** |
| HLM CLint | −0.645 | −0.096 | −0.029 | −0.012 |
| MLM CLint | 0.142 | −1.113 | 0.107 | **−0.822** |
| Caco-2 Papp A>B | 0.448 | −0.320 | 0.448 | −0.041 |
| Caco-2 Efflux | 0.402 | −0.062 | 0.456 | −0.005 |
| MPPB | 0.405 | 0.172 | 0.162 | 0.146 |
| MBPB | 0.417 | −1.231 | 0.237 | **−0.488** |

IID performance is broadly comparable between the two models. On OOD molecules, CheMeleon has a striking advantage for LogD (R² 0.444 vs 0.055) — the most well-behaved endpoint — but performs worse than XGBoost for KSOL, MLM CLint, and MBPB OOD. For endpoints where R² is negative for both models (HLM, MLM, Caco-2), neither architecture retains predictive signal on unseen series.

The main finding is consistent across both models: **OOD R² is negative for 4–5 of 8 endpoints**, confirming that the IID/OOD performance gap is a property of the problem (chemical series extrapolation is hard) rather than the model architecture. CheMeleon partially mitigates but does not resolve this failure mode.

---

## Summary

CheMeleon consistently achieves better absolute performance than XGBoost on IID evaluations, reflecting the advantage of pre-trained representations over task-specific ECFP4 + descriptors baselines. However, the core generalization findings of this study hold regardless of model:

- Performance degrades monotonically with distance from training data under cluster and time splits.
- Models fail silently on new chemical series — OOD R² is negative for the majority of high-noise endpoints.
- The magnitude of the IID→OOD performance drop is structurally similar for both architectures.

This concordance across a fingerprint-based tree model and a pre-trained GNN strengthens the paper's central claim: the observed generalization failures are features of the evaluation scenario, not artifacts of a particular modeling choice.

---

## Reference

[1] Burns, J. W., Zalte, A. S., Abreu, C. R. A., Sieg, J., Feldmann, C., Mathea, M., & Green, W. H. (2025). Deep Learning Foundation Models from Classical Molecular Descriptors. arXiv:2506.15792. https://arxiv.org/abs/2506.15792
